File: library/src/main/java/wannabe/projection/Cabinet.java
package wannabe.projection;

import wannabe.Camera;
import wannabe.Position;
import wannabe.Projected;
import wannabe.Translation;
import wannabe.Voxel;

/**
 * Renders {@link Voxel}s with a constant size, but with an offset to represent height.
 */
public class Cabinet implements Projection {
  private final Projected projected = new Projected();
  /** Horizontal offset to use for each unit of Z. */
  final int hOffsetPerZ;
  /** Vertical offset to use for each unit of Z. */
  final int vOffsetPerZ;

  /** Default constructor just has increasing Z values go up and left. */
  public Cabinet() {
    this(-1, -1);
  }

  public Cabinet(int hOffsetPerZ, int vOffsetPerZ) {
    this.hOffsetPerZ = hOffsetPerZ;
    this.vOffsetPerZ = vOffsetPerZ;
  }

  @Override public Projected project(Camera camera, Position position, int pixelSize) {
    Translation address = camera.translate(position);
    int horizOffset = hOffsetPerZ * address.z + pixelSize * address.x;
    int vertOffset = vOffsetPerZ * address.z + pixelSize * address.y;
    projected.left = camera.uiPosition.left + horizOffset;
    projected.top = camera.uiPosition.top + vertOffset;
    projected.size = pixelSize;
    projected.hDepth = -hOffsetPerZ;
    projected.vDepth = -vOffsetPerZ;

    return projected;
  }

  @Override public String toString() {
    return getClass().getSimpleName() + " " + hOffsetPerZ + ", " + vOffsetPerZ;
  }
}

File: library/src/main/java/wannabe/projection/Cabinet.kt
package wannabe.projection

import wannabe.Camera
import wannabe.Position
import wannabe.Projected

/**
 * Renders [Voxel]s with a constant size, but with an offset to represent height.
 */
class Cabinet(
  /** Horizontal offset to use for each unit of Z. */
  val hOffsetPerZ: Int,
  /** Vertical offset to use for each unit of Z.  */
  val vOffsetPerZ: Int
) : Projection {
  private val projected = Projected()

  override fun project(
    camera: Camera,
    position: Position,
    pixelSize: Int
  ): Projected {
    val (x, y, z) = camera.translate(position)
    val horizOffset = hOffsetPerZ * z + pixelSize * x
    val vertOffset = vOffsetPerZ * z + pixelSize * y
    projected.left = camera.uiPosition.left + horizOffset
    projected.top = camera.uiPosition.top + vertOffset
    projected.size = pixelSize
    projected.hDepth = -hOffsetPerZ
    projected.vDepth = -vOffsetPerZ
    return projected
  }

  override fun toString(): String = "Cabinet $hOffsetPerZ, $vOffsetPerZ"

  companion object {
    fun horizontalSequence(
      vOffset: Int,
      hStart: Int,
      hEnd: Int,
      bounce: Boolean
    ): ProjectionSequence {
      var range = (hStart..hEnd).toList()
      // Remove the last entry since it'll be added again.
      // TODO should we also remove the last entry of the combined since it would match the first?
      println("range before $range")
      if (bounce) range = range.subList(0, range.size - 1) + range.reversed()
      println("range after $range")
      return ProjectionSequence(
        projections = range.map {
          Cabinet(
            vOffsetPerZ = vOffset,
            hOffsetPerZ = it
          )
        }
      )
    }
  }
}
